Add debug output pane to cu-consolemon

AS1100K opened the pull request:
Fixes #153 

## Demo

https://github.com/user-attachments/assets/7ef668f1-3d71-4149-a15f-e3cf1a49aaae

gbin commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -535,6 +551,20 @@ impl UI {
+                let mut line = String::new();
+                if let Ok(_) = io::stdin().read_line(&mut line) {
+                    buffer.push_str(&line);
+                    *debug_output.lock().unwrap() = buffer.clone();
```
why not `*debug_output.lock().unwrap().push_str(&line)` to avoid a copy?

gbin commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -535,6 +551,20 @@ impl UI {
+                    buffer.push_str(&line);
+                    *debug_output.lock().unwrap() = buffer.clone();
+                }
+            }
```
then we need something that finds the end of lines and keep a fixed amount of logging in there.
Maybe we need a circular buffer of lines even so we don't copy or move megabytes of text from time to time (adding jitter).

gbin reviewed (commented):
Thanks, this is a good start.

makeecat commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -282,6 +288,147 @@ struct UI {
+        }
+    }
+
+    fn push_logs<A: AsRef<str>>(&self, logs: A) {
+        let mut log_lines = self.debug_log.lock().unwrap();
+        let max_col = *self.max_col.lock().unwrap();
+        log_lines.push(logs.as_ref().to_string());
+        if log_lines.len() > max_col as usize {
+            log_lines.remove(0);
+        }
+    }
```
Should we use VecDeque instead of Vec for debug_log for performance consideration?

makeecat commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -282,6 +288,147 @@ struct UI {
+
+struct LogSubscriber {
+    debug_output: Arc<DebugLog>,
+    span_registry: Arc<Mutex<HashMap<MetadataKey, Id>>>,
```
Should we consider using https://github.com/xacrimon/dashmap for concurrent hash map?

AS1100K commented:
@makeecat Thanks for the review, I have now used `VecDeque` and `DashMap`. I have also, migrated from using `Mutex` to `Channel`. I did this, as when using `Mutex` there are a lot of instances of long blocking, which makes the application feel laggy. Using Channel blocking has been prevented and thus also improves performance as now the single thread read and writes the log without attaining any lock.

Also, I think we should make this Debug pane behind feature gate as this library is using tracing [`set_global_default`](https://docs.rs/tracing/latest/tracing/subscriber/fn.set_global_default.html) which is not meant to be used by libraries as if the main crate maybe wants to send these logs or store them somewhere else but then they will not be able to use monitor.

I thinking adding a default feature like `debug-pane` would be good and will introduce flexibility for the main crate user on log behaviour.

makeecat commented:
I agree with adding a debug-panel faeture pane. In the future we may want to use tokio async to handle all IO-related operations to avoid overhead, but we can start with this implementation at the current stage.

gbin commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -282,18 +302,223 @@ struct UI {
+#[cfg(feature = "debug_pane")]
+struct DebugLog {
+    debug_log: VecDeque<String>,
+    max_rows: Arc<Mutex<u16>>,
```
can we get away with a simple Mutex as we only protect a u16? Arc is usually when you want to ship the reference around. Even better an AtomicU16 :)

gbin commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -282,18 +302,223 @@ struct UI {
+struct LogSubscriber {
+    tx: SyncSender<String>,
+    span_registry: Arc<DashMap<MetadataKey, Id>>,
+    next_id: Arc<Mutex<u64>>,
```
same here

gbin commented on components/monitors/cu_consolemon/Cargo.toml:
```diff
@@ -23,3 +28,7 @@ tui-nodes = "0.8"
 tui-widgets = "0.4"  # 0.4.0 brings ratatui 0.29
 color-eyre = "0.6"
 gag = "1.0.0"
+tracing = { version = "0.1", optional = true }
```
candid question: why do we need tracing? 

This adds another logging facility aside the one from Copper itself [cu29-log](https://www.copper-robotics.com/whats-new/deep-dive-2-structured-text-logging) where we can hook up the output of our structured logging to the standard [log crate](https://github.com/copper-project/copper-rs/blob/master/core/cu29_log_runtime/src/lib.rs#L132) if the user really wants live text logs (only in debug + a flag).

^ why not splitting it in 2, one that subscribed to the log we already use in Copper and an stderr/stdout capture.

Maybe to make it fancy: clearly mark in the ui either with a color or an icon from where the line is coming from (so the user gets a hint if this is a dependency going to std or the standard rust logging)

then the pane is just a list of ANSI Strings to render from the bottom.

gbin replied:
BTW: if we want to also support tracing as an output of the structured logging in Copper, that is another story, I would be supportive as it brings value to users who are already using tracing.

AS1100K replied:
> stdout capture

The problem is we can't listen to `stdout` as locking `stdin` would make the entire terminal irresponsive and will also capture `stdout` send by `crossterm` which will mess the entire stuff. Therefore, I went by adding support for `tracing` and `log` _(via `tracing-log`)_ crate as both `tracing` and `log` are very popular logging frameworks and this will allow any other libraries logs to be seen by the end user.

gbin replied:
yeah but we don't want tracing on the main stack, remember Copper is 0 allocation after the tasks area starting, we don't even construct Strings. This is why we implemented the structured logging.

gbin commented:
Thanks @AS1100K! My biggest question here is why do we want to introduce tracing in the middle of it. Also, I don't see where we keep the buffer from not growing indefinitely, we should have *somewhere* a constant or parameter saying: we keep the last 1000 lines of back buffer. Otherwise your robot will run out of memory at some point :)

gbin commented on components/monitors/cu_consolemon/Cargo.toml:
```diff
@@ -10,6 +10,11 @@ categories.workspace = true
 repository.workspace = true
 
+[features]
+default = ["debug_pane"]
```
nice call here!

gbin commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -570,16 +823,48 @@ impl CuMonitor for CuConsoleMon {
+                ui.run_app(&mut terminal).expect("Failed to run app");
+            }
+
+            #[cfg(not(feature = "debug_pane"))]
```
```suggestion
            #[cfg(not(feature = "debug_pane"))]
            {
```

gbin replied:
Just a simple scope would work no?

gbin commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -570,16 +823,48 @@ impl CuMonitor for CuConsoleMon {
             // restoring the terminal
-            drop(print_gag);
+            #[cfg(not(feature = "debug_pane"))]
+            drop(stderr_gag);
```
```suggestion
           }
```

AS1100K commented:
> Also, I don't see where we keep the buffer from not growing indefinitely

I also kept that in mind, and only keep the number of logs according to terminal height. here https://github.com/copper-project/copper-rs/pull/219/files#diff-77cbd66ea06743fe2a1e13a87d5525ab87330d6cc6d548f35068b256168b611cR352-R356

gbin commented:
> > Also, I don't see where we keep the buffer from not growing indefinitely
> 
> I also kept that in mind, and only keep the number of logs according to terminal height. here https://github.com/copper-project/copper-rs/pull/219/files#diff-77cbd66ea06743fe2a1e13a87d5525ab87330d6cc6d548f35068b256168b611cR352-R356

aaaah... ok ... we need a back buffer :) people will need to scroll back up a little bit

AS1100K commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -282,18 +302,223 @@ struct UI {
+}
+
+#[cfg(feature = "debug_pane")]
+impl Subscriber for LogSubscriber {
```
> This adds a lot of complexity

For most of the stuff we could just have `event` function and have other functions `unimplemented!()` but this would become a issue if some libraries are using `span` macro as the other functions in the trait are just to support [span](https://docs.rs/tracing/latest/tracing/span/index.html) events

AS1100K commented:
@gbin Thanks for you review, I have now removed the support for `tracing` crate and now it supports `cu29-log` and `log` crate. Also, while doing so I introduced a new argument in [`basic_copper_setup`](https://github.com/copper-project/copper-rs/compare/master...AS1100K:copper-rs:add-debug-output#diff-bc7e34b15032b0327094bade29a60c828a1256f7bc561297a6046acf3e8163d9R31) and [`LoggerRuntime::init`](https://github.com/copper-project/copper-rs/compare/master...AS1100K:copper-rs:add-debug-output#diff-6356a806527043ab67b7bf477ac642af378e7e4f1a632be69e07b54729239e23R61) function. I would love some feedback here as I am not sure if this is a right call here.



> > > Also, I don't see where we keep the buffer from not growing indefinitely
> > 
> > 
> > I also kept that in mind, and only keep the number of logs according to terminal height. here https://github.com/copper-project/copper-rs/pull/219/files#diff-77cbd66ea06743fe2a1e13a87d5525ab87330d6cc6d548f35068b256168b611cR352-R356
> 
> aaaah... ok ... we need a back buffer :) people will need to scroll back up a little bit

Also, I would also implement it soon

gbin commented:
Cool! Thank you for your patience!

cu-consolemon is not part of core, I don't think it should: for example in production robots should not have it and might use a different implementation behind the Monitoring API.

what about hooking that up either as a Log destination (that will be picked up when consolemon is in) or, another possible way is to add a logging API to the monitoring API (that consolemon relies on and decouples it from core)?

gbin commented on components/monitors/cu_consolemon/src/lib.rs:
```diff
@@ -282,18 +298,141 @@ struct UI {
+    debug_output: DebugLog,
+}
+
+#[cfg(feature = "debug_pane")]
```
put all that in a module and make a use of the module when the feature is on

gbin commented on core/cu29_log_runtime/Cargo.toml:
```diff
@@ -21,6 +21,7 @@ cu29-clock = { workspace = true }
 bincode = { workspace = true }
 smallvec = "1.13.2"
 log = "0.4.22"
+once_cell = "1.20"
```
this is not integrated to the stdlib in the recent Rust versions?

gbin reviewed (approved):
Ok, 2 comments left and we merge this.

gbin commented:
Thanks @AS1100K good work!
